installinator/src/reporter.rs

```diff
@@ -49,129 +78,418 @@ impl ProgressReporter {
+                    Ok(()) => peers_rx.borrow_and_update(),
+                    Err(error) => {
+                        slog::warn!(
+                            log,
+                            "failed to receive peers (sender dropped, \
+                             discovery task likely completed), exiting: {}",
+                            error
+                        );
+                        return;
+                    }
+                };
+
+                let Some(peers) = peers.as_ref() else {
+                    continue;
+                };
+
+                let before_peers =
+                    peer_tasks.keys().cloned().collect::<BTreeSet<_>>();
+                let after_peers = peers.peers().clone();
+                let diff = before_peers.diff(&after_peers);
+
+                // For each peer in diff.removed, cancel the report task.
+                for peer in diff.removed {
+                    let task = peer_tasks
+                        .remove(&peer)
+                        .expect("present in removed => present in peer_tasks");
+                    _ = task.cancel_tx.cancel(());
                 }
-            };
 
-            // We could use StreamExt::any() here, but we're choosing to avoid
-            // short-circuiting for now in favor of reaching out to all peers.
-            //
-            // TODO: is this correct? If a peer takes too long this might end up
-            // stalling updates for everyone -- feels wrong. The tradeoff is
-            // that if two servers both say, only one of them will
-            // deterministically get the update. Need to decide post-PVT1.
-            let last_reported = report.last_seen;
-            let results: Vec<_> =
-                futures::stream::iter(peers.peers().iter().copied())
-                    .map(|peer| {
-                        report_backend.send_report_to_peer(
-                            peer,
-                            update_id,
-                            report.clone(),
-                        )
-                    })
-                    .buffer_unordered(8)
-                    .collect()
-                    .await;
-
-            if results.iter().any(|res| res.is_ok()) {
-                Some(last_reported)
-            } else {
-                None
+                // For each peer in diff.added, spawn a new report task.
+                for peer in diff.added {
+                    let task = ReportTask::new(
+                        *peer,
+                        update_id,
+                        // Note: report_rx.clone() starts off as seen. However,
+                        // the initial state of report_rx is None, so if there's
+                        // an report in the receiver, it will always be treated
+                        // as unseen. This means that if a report is available,
+                        // the task will immediately dispatch it to the peer.
```

Comment:
Just a general question on watch channel style: to avoid having to think about this, would it make sense to structure consumers of a watch channel as something like

```rust
loop {
    let current_value = rx.borrow_and_update();

    // ... do whatever with current value ...

    match rx.changed().await {
        Ok(()) => continue,
        Err(_) => break,
    }
}
```

instead of blocking on `changed()` first?

Reply:
I think that's totally reasonable if the initial value in the watch channel is meaningful. In our case it is `None` so I decided to do it this way. Happy to change to your suggestion if it makes more sense to you, though -- what do you think?

Reply:
Yeah I think what you have here is fine :+1: